File: 1. Study/0603 - 효율성(슬라이딩윈도우, 해쉬)/kimshin/1.최대매출.py
def sol(k,arr):
    answer=0
    sum=0
    for i in range(k):
        sum += arr[i]
    answer=sum
    for j in range(k,len(arr)):
        sum += arr[j] - arr[j-k]
        answer = max(answer,sum)
    return answer


a=[12,15,11,20,25,10,20,19,13,15]
print(sol(3,a))

File: 1. Study/0603 - 효율성(슬라이딩윈도우, 해쉬)/mojaeya/1_최대매출.js
let arr = [12, 15, 11, 20, 25, 10, 20, 19, 13, 15];

console.log(solution(10, 3, arr));

// CODE 1
function solution(N, K, sale) {
  let answer = 0;
  let curSum = 0;

  for (let i = 0; i < K; i++) {
    curSum += sale[i];
    answer = curSum;
  }

  for (let j = K; j < N; j++) {
    curSum += sale[j] - sale[j - K];
    answer = Math.max(answer, curSum);
  }

  return answer;
}

// CODE 2
function solution(n, k, arr) {
  let lt = (rt = currentSum = 0);
  let max = Number.MIN_SAFE_INTEGER;

  while (rt < n) {
    currentSum += arr[rt];
    if (rt - lt + 1 === k) {
      max = Math.max(max, currentSum);
      currentSum -= arr[lt++];
    }
    rt++;
  }
  return max;
}
